### ISSUE TITLE
Postgres planner couldn't plan query with non-select CTE sharing

### ISSUE THREAD

**Issue by KnightMurloc** [greenplum-db/gpdb#16776]
## Bug Report
When CTE subquery contains modifying DML operation then postgres planner always tried to inline it to upper query even if the upper query has multiple links to CTE subquery and inlining leads to multiple modifications. Workaround is to enable `gp_cte_sharing`. But in this case, the planning fails. The reason for this is that when the planner was preparing shared scan over CTE, that contained modifying operation, the Materialize node did not retrieved the target list from ModifyTable node. That led to the init of Materialize node was failed with the error: `"Material operator does not support projection"`, because of that the number of elements in its target list and the child node are different. It also led to Shared Scan node also doesn't had target list. This happened because ModifyTable node don't set its own target list on creation.

I have a patch (https://github.com/arenadata/gpdb/commit/3efffb4194972f973e70642220357e46afff9097) that solves this by sets target list of Materialize node on it's creation from returning lists of child ModifyTable node. We pick first returning list because the first one (in the case of inherited tables) contains the target list of the root table while the rest of the child tables. We can take the returning list before it processing in `set_plan_refs` because Materialize node has its own processing in `set_plan_refs`. Returning list is processed in `set_returning_clause_references` which handles situations with outer Vars. Since ModifyTable does the projection, we don't need to handle these situations.
### Greenplum version or build
`PostgreSQL 9.4.26 (Greenplum Database 6.26.0+dev.8.gf451b4d1fe build dev) on x86_64-unknown-linux-gnu, compiled by gcc (Ubuntu 11.3.0-1ubuntu1~22.04.1) 11.3.0, 64-bit compiled on Nov 23 2023 15:22:55 (with assert checking)`
### OS version and uname -a
Ubuntu 22.04 LTS (docker)
`Linux MURLOC 6.1.55-1-MANJARO #1 SMP PREEMPT_DYNAMIC Sat Sep 23 12:13:56 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux`
### autoconf options used ( config.status --config )
`'--enable-debug-extensions' '--with-python' '--disable-gpcloud' '--enable-depend' '--enable-cassert' 'CFLAGS=-O0 -g3 -ggdb -fno-omit-frame-pointer -Wno-error=deprecated-copy' 'CXXFLAGS=-O0 -g3 -fno-omit-frame-pointer -Wno-error=deprecated-copy' '--prefix=/gpdb/gpdb_bin' '--with-libxml`
### Installation information ( pg_config )
```
BINDIR = /gpdb/gpdb_bin/bin
DOCDIR = /gpdb/gpdb_bin/share/doc/postgresql
HTMLDIR = /gpdb/gpdb_bin/share/doc/postgresql
INCLUDEDIR = /gpdb/gpdb_bin/include
PKGINCLUDEDIR = /gpdb/gpdb_bin/include/postgresql
INCLUDEDIR-SERVER = /gpdb/gpdb_bin/include/postgresql/server
LIBDIR = /gpdb/gpdb_bin/lib
PKGLIBDIR = /gpdb/gpdb_bin/lib/postgresql
LOCALEDIR = /gpdb/gpdb_bin/share/locale
MANDIR = /gpdb/gpdb_bin/share/man
SHAREDIR = /gpdb/gpdb_bin/share/postgresql
SYSCONFDIR = /gpdb/gpdb_bin/etc/postgresql
PGXS = /gpdb/gpdb_bin/lib/postgresql/pgxs/src/makefiles/pgxs.mk
CONFIGURE = '--enable-debug-extensions' '--with-python' '--disable-gpcloud' '--enable-depend' '--enable-cassert' 'CFLAGS=-O0 -g3 -ggdb -fno-omit-frame-pointer -Wno-error=deprecated-copy' 'CXXFLAGS=-O0 -g3 -fno-omit-frame-pointer -Wno-error=deprecated-copy' '--prefix=/gpdb/gpdb_bin' '--with-libxml'
CC = gcc
CPPFLAGS = -D_GNU_SOURCE -I/usr/include/libxml2 -I/gpdb/gpdb_bin/include
CFLAGS = -Wall -Wmissing-prototypes -Wpointer-arith -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -fno-aggressive-loop-optimizations -Wno-unused-but-set-variable -Wno-address -Werror=implicit-fallthrough=3 -Wno-format-truncation -Wno-stringop-truncation -O0 -g3 -ggdb -fno-omit-frame-pointer -Wno-error=deprecated-copy -std=gnu99 -Werror=uninitialized -Werror=implicit-function-declaration -I/gpdb/gpdb_bin/include
CFLAGS_SL = -fPIC
LDFLAGS = -Wl,--as-needed -Wl,-rpath,'/gpdb/gpdb_bin/lib',--enable-new-dtags -L/gpdb/gpdb_bin/lib
LDFLAGS_EX = 
LDFLAGS_SL = 
LIBS = -lpgcommon -lpgport -lxerces-c -lxml2 -lrt -lzstd -lcrypt -lm -L/gpdb/gpdb_bin/lib
VERSION = PostgreSQL 9.4.26
```
### Expected behavior
successful query planning
### Actual behavior
`ERROR:  Material operator does not support projection (nodeMaterial.c:358)`
### Step to reproduce the behavior
```
set gp_cte_sharing to on;
create table t1 (
    c1 int,
    c2 varchar(100))
distributed randomly;
explain with cte1 as (insert into t1 values (1,2) returning *) select * from cte1 a join cte1 b using(c1);
```

**Comment by Aegeaner** [comment]
@KnightMurloc GP7 does not have this issue, even without `make_material()` modified, do you know why? We should try to keep align with upstream when possible.

**Comment by KnightMurloc** [comment]
in GP7 we don't use the Material node with Shared Scan so we don't get an error.

**Comment by Aegeaner** [comment]
> in GP7 we don't use the Material node with Shared Scan so we don't get an error.

Right, the implementation of Shared Scan have changed.

**Comment by KnightMurloc** [comment]
yes, it looks like GP7 is not affected by the problem. but for GP6, the problem is relevant.
